File: array/844_Backspace_String_Compare/844.java
class Solution {
    public boolean backspaceCompare(String s, String t) {
        int i = s.length() - 1;
        int j = t.length() - 1;
        while (i >= 0 || j >= 0) {
            int skipS = 0;
            while (i >= 0 && (s.charAt(i) == '#' || skipS > 0)) {
                if (s.charAt(i) == '#') {
                    skipS += 1;
                } else {
                    skipS -= 1;
                }
                i -= 1;
            }

            int skipT = 0;
            while (j >= 0 && (t.charAt(j) == '#' || skipT > 0)) {
                if (t.charAt(j) == '#') {
                    skipT += 1;
                } else {
                    skipT -= 1;
                }
                j -= 1;
            }

            if (i >= 0 && j >= 0 && s.charAt(i) != t.charAt(j)) {
                return false;
            }

            if ((i >= 0) != (j >= 0)) {
                return false;
            }

            i -= 1;
            j -= 1;
        }
        return true;
    }
}

File: array/844_Backspace_String_Compare/844.py
class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        def next_valid_char_index(s, index):
            skip = 0
            while index >= 0:
                if s[index] == '#':
                    skip += 1
                elif skip > 0:
                    skip -= 1
                else:
                    return index
                index -= 1
            return index

        i, j = len(s) - 1, len(t) - 1
        while i >= 0 or j >= 0:
            i = next_valid_char_index(s, i)
            j = next_valid_char_index(t, j)

            if i >= 0 and j >= 0 and s[i] != t[j]:
                return False
            if (i >= 0) != (j >= 0):
                return False

            i -= 1
            j -= 1
        return True
